### PR TITLE
Altframe hooks

### PR DESCRIPTION
**Pull request by QuPsi**
Introduces ability to set the fence max altitude frame. Also includes @peterbarker's param-validate functionality.

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Looks like you are reverting some changes from master by accident. My guess is that a couple of merge conflicts went south.
You should fix that first

File: libraries/AP_Common/Location.cpp
Comment (peterbarker) [inline libraries/AP_Common/Location.cpp:116]:
```suggestion
bool Location::convert_altitude_frame(AltFrame current_frame, AltFrame desired_frame, int32_t alt_in_cm, int32_t &alt_out_cm) const
```

File: libraries/AP_Common/Location.h
Comment (peterbarker) [inline libraries/AP_Common/Location.h:47]: This should be static, yes?

File: libraries/AP_Param/AP_Param.cpp
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.cpp:39]: Why is this included?

File: ArduCopter/system.cpp
Comment (peterbarker) [inline ArduCopter/system.cpp:18]: This would be better done as a `switch`, enumerating the valid values.

File: ArduCopter/system.cpp
Comment (peterbarker) [inline ArduCopter/system.cpp:20]: Pointless `else`

File: ArduCopter/system.cpp
Comment (peterbarker) [inline ArduCopter/system.cpp:21]: These validation routines were never meant to modify values - only validate them.

Coupling parmeters like this is going to lead to all sorts of confusion.  Pretty sure we want to leave it up to the user to set both the frame and altitude at the same time.

File: ArduPlane/Parameters.cpp
Comment (peterbarker) [inline ArduPlane/Parameters.cpp:970]: These are the changes you're reverting when you shouldn't be.

File: libraries/AC_Fence/AC_Fence.h
Comment (peterbarker) [inline libraries/AC_Fence/AC_Fence.h:48]: This definitely doesn't look right.  What happened to people's persistence here?!

File: libraries/AP_Param/AP_Param.h
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.h:67]: Bad whitespace change.

File: libraries/AP_Param/AP_Param.h
Comment (peterbarker) [inline libraries/AP_Param/AP_Param.h:27]: Why?

File: libraries/GCS_MAVLink/GCS_Param.cpp
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Param.cpp:74]: Another master change reverted.  Not good.

**Review by peterbarker — CHANGES_REQUESTED** [review]
I can see some uses, but I am curious - what is your need to be able to specify the altitude frame of the altitude fences?

**Comment by amilcarlucas** [comment]
Do not merge master into this branch. Instead rebase this branch on top of master.
Instructions are here: https://ardupilot.org/dev/docs/git-interactive-rebase.html

**Comment by peterbarker** [comment]
Closing in favour of https://github.com/ArduPilot/ardupilot/pull/18038

**Pull request closed without merging**
